python_modules/libraries/dagster-components/dagster_components/resolved/base.py

```diff
@@ -0,0 +1,269 @@
+
+from dagster import _check as check
+from dagster._utils.pydantic_yaml import _parse_and_populate_model_with_annotated_errors
+from dagster_shared.yaml_utils import parse_yaml_with_source_positions
+from pydantic import BaseModel, PydanticSchemaGenerationError, create_model
+from pydantic.fields import Field, FieldInfo
+from typing_extensions import TypeGuard
+
+from dagster_components.resolved.context import ResolutionContext
+from dagster_components.resolved.errors import ResolutionException
+from dagster_components.resolved.model import (
+    ResolvableModel,
+    Resolver,
+    _is_implicitly_resolved_type,
+    derive_field_resolver,
+)
+
+try:
+    # this type only exists in python 3.10+
+    from types import UnionType  # type: ignore
+except ImportError:
+    UnionType = Union
+
+
+class _TypeContainer(Enum):
+    SEQUENCE = auto()
+    OPTIONAL = auto()
+
+
+_DERIVED_MODEL_REGISTRY = {}
+
+
+class Resolvable:
+    """This base class makes something able to "resolve" from yaml.
+
+    This is done by:
+    1) Deriving a pydantic model to provide as schema for the yaml.
+    2) Resolving an instance of the class by recursing over an instance
+        of the derived model loaded from schema compliant yaml and
+        evaluating any template strings.
+
+    The fields/__init__ arguments of the class can be Annotated with
+    Resolver to customize the resolution or model derivation.
+    """
+
+    @classmethod
+    def model(cls) -> type[ResolvableModel]:
+        return derive_model_type(cls)
+
+    @classmethod
+    def resolve_from_model(cls, context: "ResolutionContext", model: ResolvableModel):
+        return cls(**_resolve_fields(model, cls, context))
+
+    @classmethod
+    def resolve_from_yaml(cls, yaml: str):
+        parsed_and_src_tree = parse_yaml_with_source_positions(yaml)
+        model = _parse_and_populate_model_with_annotated_errors(
+            cls=cls.model(),
+            obj_parse_root=parsed_and_src_tree,
+            obj_key_path_prefix=[],
+        )
+        # support adding scopes?
+        context = ResolutionContext.default(parsed_and_src_tree.source_position_tree)
+        return cls.resolve_from_model(context, model)
+
+
+class _Unset: ...
+
+
+def derive_model_type(
+    target_type: type[Resolvable],
+) -> type[ResolvableModel]:
+    if target_type not in _DERIVED_MODEL_REGISTRY:
+        model_name = f"{target_type.__name__}Model"
+
+        model_fields: dict[
+            str, Any
+        ] = {}  # use Any to appease type checker when **-ing in to create_model
+
+        for name, (annotation, has_default) in _get_annotations(target_type).items():
+            field_resolver = _get_resolver(annotation, name)
+            field_name = field_resolver.model_field_name or name
+            field_type = field_resolver.model_field_type or annotation
+
+            field_infos = []
+            if has_default:
+                field_infos.append(Field(default=_Unset))
```

Comment:
what's the deal with _Unset here?

Reply:
so the line of thought here is instead of trying to accurately propagate defaults which can be not just values  but also factory functions, we just use a special marker value `_Unset` for the default value and then filter that out before passing kwargs through [1] . Then since we are omitting those kwargs that will kick in the real defaulting on the target class 

ill add comment in code too

Reply:
perfect, that makes sense